doc/performance-investigations/0013-gunicorn-worker-thread-config/summary.md

```diff
@@ -0,0 +1,112 @@
+| 9 x 10          | 67              | 120             | 140             | 200             | 300               | 580              | 172                     | 214                      | 193                             | 326,015        |
+| 10 x 7          | 65              | 120             | 140             | 200             | 310               | 660              | 172                     | 212                      | 192                             | 326,886        |
+| 10 x 10         | 73              | 130             | 160             | 230             | 340               | 620              | 196                     | 246                      | 221                             | 323,284        |
+
+- There were 0 failures for these configurations.
+
+### Performance comparison of thread counts
+
+The following results were omitted because the 99th percentile response was above 500ms:
+1. All workers with a thread count of 1
+2. All permutations of 1 and 2 workers.
+
+#### Workers: 3
+
+![](images/performance_graph_workers3.png)
+
+#### Workers: 4
+
+![](images/performance_graph_workers4.png)
+
+#### Workers: 5
+
+![](images/performance_graph_workers5.png)
+
+#### Workers: 6
+
+![](images/performance_graph_workers6.png)
+
+#### Workers: 7
+
+![](images/performance_graph_workers7.png)
+
+#### Workers: 8
+
+![](images/performance_graph_workers8.png)
+
+#### Workers: 9
+
+![](images/performance_graph_workers9.png)
+
+#### Workers: 10
+
+![](images/performance_graph_workers10.png)
+
+#### Comparison of the 99th percentile between worker and thread counts
+
+![](images/performance_graph_workers_by_threads.png)
+
+### Comparison of the 99th percentile between workers with a thread count of 2
+
+The performance of the app does not seem to get noticeably better past 2 threads. The graph below highlights the performance of different workers under 2 threads over 3 runs.
+
+![](images/performance_graph_workers_by_2_threads_99th_avg.png)
+
+## Observations
+
+- Using a thread count of 1 or worker count of 1 or 2 is not feasible.
+- Our current configuration of 7 workers by 7 threads is not the most optimal.
+- Under this test scenario, having more than 2 threads did not have any noticeable advantages in response times.
+    - Along with this, having more than 4 workers does not seem to have any significant improvement in response times.
+
+## Decision
+- Since different configuration has different pros and cons for different percentiles a few different worker/thread configurations will need further testing.
+- Run a small stress test to investigate the impact of thread counts when the app is under stress.
+    - Stress the app while using 7 workers with 7 threads vs 7 workers with 2 threads until the 99th percentile reaches 1 second. Following on from that, run a small stress test with reduced number of workers.
```

Comment:
I don't think the first point is a decision, rather it leads into the decisions that we should run some stress tests:
```suggestion
- Since different configurations have different pros and cons for different percentiles, a few different worker/thread configurations should be tested further in small scale stress tests:
  - Compare a baseline of 7 workers with 7 threads vs 7 workers with 2 threads, until the 99th percentile reaches 1 second. 
  - Following on from that, run a test with a reduced number of workers.
```